File: leetcode/hard/1203_sortItems.ts
function sortItems(
    n: number,
    m: number,
    group: number[],
    beforeItems: number[][],
): number[] {
    return [];
}

/*

n: 8, m: 2,
group: [
0:  -1,
1:  -1,
2:  1,
3:  0,
4:  0,
5:  1,
6:  0,
7:  -1
],
beforeItems: [
0:  [],
1:  [6],
2:  [5],
3:  [6],
4:  [3, 6],
5:  [],
6:  [],
7:  []
],

-1: 0, 1, 7
 0: 3, 4, 6,
 1: 2, 5,

0, 7, 5, 2, 6, 3, 4, 1

---

n: 8, m: 2,
group: [
0:  -1,
1:  -1,
2:  1,
3:  0,
4:  0,
5:  1,
6:  0,
7:  -1
],
beforeItems: [
0:  [],
1:  [6],
2:  [5],
3:  [6],
4:  [3],
5:  [],
6:  [4],
7:  []
],

-1: 0, 1, 7
 0: 3, 4, 6
 1: 2, 5

0, 7, 5, 2, ...

*/

const test = () => {
    const params = [
        {
            input: {
                n: 8,
                m: 2,
                group: [-1, -1, 1, 0, 0, 1, 0, -1],
                beforeItems: [[], [6], [5], [6], [3, 6], [], [], []],
            },
            output: [6, 3, 4, 1, 5, 2, 0, 7],
        },
        {
            input: {
                n: 8,
                m: 2,
                group: [-1, -1, 1, 0, 0, 1, 0, -1],
                beforeItems: [[], [6], [5], [6], [3], [], [4], []],
            },
            output: [],
        },
    ];

    params.forEach(({ input, output }) => {
        const { n, m, group, beforeItems } = input;
        const result = sortItems(n, m, group, beforeItems);

        console.log(
            JSON.stringify(result) === JSON.stringify(output)
                ? "SUCCESS "
                : "ERROR ",
            "input",
            JSON.stringify(input),
            "output",
            output,
            "result",
            result,
        );
    });
};

test();

File: leetcode/hard/1203_sortItems.py
import sys
sys.setrecursionlimit(1500)
from typing import List
import json
from collections import deque, defaultdict
import heapq

class Solution:
    def sortItems(self, n: int, m: int, group: List[int], beforeItems: List[List[int]]) -> List[int]:
        res = []

        myGroups = defaultdict(list)
        childrenForGroups = defaultdict(set)
        parentsForGroups = defaultdict(set)
        childrenForNodes = [0]*n
        parentsForNodes = [[] for i in range(n)]

        for pr in range(n):
            prKey = str(group[pr]) if group[pr] != -1 else 's_' + str(pr)
            myGroups[prKey].append(pr)

            for ch in beforeItems[pr]:
                chKey = str(group[ch]) if group[ch] != -1 else 's_' + str(ch)
                if chKey != prKey:
                    parentsForGroups[chKey].add(prKey)
                    childrenForGroups[prKey].add(chKey)
                else:
                    parentsForNodes[ch].append(pr)
                    childrenForNodes[pr] += 1


        groupsQueu = deque()

        for gr in myGroups:
            if not gr in childrenForGroups or len(childrenForGroups[gr]) == 0:
                groupsQueu.append(gr)

        while groupsQueu:
            gr = groupsQueu.popleft()

            items = []
            itemsQueue = deque()
            for node in myGroups[gr]:
                if childrenForNodes[node] == 0:
                    itemsQueue.append(node)

            while itemsQueue:
                item = itemsQueue.popleft()
                items.append(item)
                for pr in parentsForNodes[item]:
                    childrenForNodes[pr] -= 1
                    if childrenForNodes[pr] == 0:
                        itemsQueue.append(pr)

            if len(items) != len(myGroups[gr]):
                return []
            else:
                res += items

            for prGr in parentsForGroups[gr]:
                childrenForGroups[prGr].remove(gr)
                if len(childrenForGroups[prGr]) == 0:
                    groupsQueu.append(prGr)


        return res if len(res) == n else []


'''
-1:
0, 1: 5-1 6-0, 7

0:
3 6-0, 4 3-0, 6-0,6

1: 0
2 5-1, 5 3-0,


'''


def test ():
    params = [
        {
            'input': [
                8, 2, [-1,-1,1,0,0,1,0,-1], [[],[6],[5],[6],[3,6],[],[],[]],
            ],
            'output': [6,3,4,1,5,2,0,7],
        },
        {
            'input': [
                8, 2, [-1,-1,1,0,0,1,0,-1], [[],[6],[5],[6],[3],[],[4],[]],
            ],
            'output': [],
        },
        #
        {
            'input': [
                8, 2, [-1,-1,1,0,0,1,0,-1], [[],[5, 6],[5],[6],[3,6],[3],[],[]],
            ],
            'output': [6,3,4,5,2,1,0,7],
        },
    ]
    solution = Solution()

    for param in params:
        n, m, group, beforeItems = param['input']
        result = solution.sortItems(n, m, group, beforeItems)

        correct = json.dumps(result) == json.dumps(param['output'])

        msg = 'SUCCESS' if correct else 'ERROR'
        msg += '\n'
        if not correct:
            msg += 'input ' + json.dumps(param['input']) + '\n'
            msg += 'output ' + json.dumps(param['output']) + '\n'
            msg += 'result ' + json.dumps(result) + '\n'

        print(msg)


if __name__ == '__main__':
    test()
